src/main/java/dev/jbang/cli/BaseScriptCommand.java

```diff
@@ -303,13 +309,20 @@ private static ScriptResource fetchFromURL(String scriptURL) {
 					throw new ExitException(10, exmsg);
 			}
 
-			scriptURL = swizzleURL(scriptURL);
-
-			String urlHash = Util.getStableID(scriptURL);
-			File urlCache = Settings.getCacheDir(Settings.CacheClass.urls).resolve(urlHash).toFile();
-			Path path = Util.downloadFileSwizzled(scriptURL, urlCache);
-
-			return new ScriptResource(scriptURL, urlCache, path.toFile());
+			if (Util.isGistURL(scriptURL)) {
+				GistCollection gistCollection = Util.extractFileFromGistWithSources(scriptURL);
```

Comment:
is there a reason why you don't put this logic into swizzleUrl() ?

Reply:
`swizzleUrl()` returns a String, but I need more than that, otherwise I would lose the information of what was already done, like source files path.

The way jbang works today without this PR, it gets the URL and then tries to download the files, but that's redundant work.

What does this PR solve?
  1. Make Gist work when there's a `#` in the URL, eg: jbang https://gist.github.com/joshlong/d8768b697a73b62bf42d8b74ec43b760#file-greetingsserver-java
  2. Make Gist work for multiple files, eg (3 files): jbang https://gist.github.com/tivrfoa/8e6ea001f168fd4ef14763ceca3e5ab6
  3. Fix: it considered the first file as the main file, but that's not necessarily the case. I check which file contains the `main` method.
  4. Fix: it redundantly downloaded the files, but that's not necessary for Gist because the content of all files are already returned by the Github API.

Reply:
Yes, so 1, 3 and 4 I'm fine with. its 2 that makes things inconsistent. can you disable that and make PR for it working generically ? 

btw. for 4, a gist will not necessarily have *all* files as you can technically have a ref between two gists or just any url thus multiple requests is needed (which can then be cached) - at least when fully implemented.

Reply:
> Yes, so 1, 3 and 4 I'm fine with. its 2 that makes things inconsistent. can you disable that and make PR for it working generically ?

By *inconsistent* you mean that multiple sources was only working for Gists?
I changed the code. I think it's now working for everything.

> 
> btw. for 4, a gist will not necessarily have _all_ files as you can technically have a ref between two gists or just any url thus multiple requests is needed (which can then be cached) - at least when fully implemented.

I didn't know you could have url in //SOURCES. That's interesting.
My new change handles that.

Now the followings also work:
  1. Gist work when there's a # in the URL, eg: jbang https://gist.github.com/joshlong/d8768b697a73b62bf42d8b74ec43b760#file-greetingsserver-java
  2. Multiple sources in Gist: jbang https://gist.github.com/tivrfoa/8e6ea001f168fd4ef14763ceca3e5ab6
  3. Multiple sources in files one -> two -> three:
```
./jbang/bin/jbang --verbose ~/dev/java/jbang/one.java
[jbang] System Java version detected as 11
[jbang] System Java version matches requested version 11
[jbang] Building jar...
[jbang] compile: /usr/lib/jvm/java-11-openjdk-amd64/bin/javac -d /home/leandro/.jbang/cache/jars/one.java.1fc5dcbe49ea9d83719ece10a2fb33b09d91a76cb883e660a4df77b03a8e3c89 /home/leandro/dev/java/jbang/one.java /home/leandro/dev/java/jbang/two.java /home/leandro/dev/java/jbang/three.java
[jbang] System Java version matches requested version 11
[jbang] run: /usr/lib/jvm/java-11-openjdk-amd64/bin/java -classpath /home/leandro/.jbang/cache/jars/one.java.1fc5dcbe49ea9d83719ece10a2fb33b09d91a76cb883e660a4df77b03a8e3c89.jar: one
two constructor
three constructor
```

I changed the test `multiSourcesHttp` so now `two` references (//SOURCES) `three`.

One thing that already worked but needs a test: `testFetchFromGistReferencingAnotherURL`
I tried to add a test for Gist Referencing URL in //SOURCE, but it doesn't work because of the "trust" checking.
But it's working if you run: jbang https://gist.github.com/tivrfoa/39a0dee0ef32a75a064fe9c59c2bd68a

Reply:
Hi @maxandersen ,

Now the PR is much simpler.

I was able to simplify it by analyzing the Gist API, eg:
https://api.github.com/gists/8e6ea001f168fd4ef14763ceca3e5ab6

So the problem was that JBang sees:<br>
raw_url	ht<span>tp</span>s://gist.githubusercontent.com/tivrfoa/8e6ea001f168fd4ef14763ceca3e5ab6/raw/**5d4fb921c76c686bc96a8bf4993304c2d95dcbab**/one.java

And then tries the same path for other files, eg:
ht<span>tp</span>s://gist.githubusercontent.com/tivrfoa/8e6ea001f168fd4ef14763ceca3e5ab6/raw/**5d4fb921c76c686bc96a8bf4993304c2d95dcbab**/two.java

But this path **does not exist.** The path for `two.java` is:
raw_url	ht<span>tp</span>s://gist.githubusercontent.com/tivrfoa/8e6ea001f168fd4ef14763ceca3e5ab6/raw/**dfc4ac2bf9ad11cd8db5f2b151162bcd3b11efda**/two.java

But then I found out that if we use history[0].version, then it's the same path for all files! :smiley:

So I basically had to make a change in swizzleURL for Gist: `extractFileFromGist`

```diff
+				String[] tmp = entry.getValue().get("raw_url").split("/raw/");
+				String prefix = tmp[0] + "/raw/";
+				String suffix = tmp[1].split("/")[1];
+				String mostRecentVersionRawUrl = prefix + gist.history[0].version + "/" + suffix;


	static class Gist {
		Map<String, Map<String, String>> files;
+		History[] history;
+	}
+
+	static class History {
+		String version;
+	}
```

Reply:
I still need to handle more use cases:

  1. Local file referencing a url
```
leandro:distributions$ ./jbang/bin/jbang --verbose ~/dev/java/jbang/Main.java 
[jbang] [ERROR] /home/leandro/dev/java/jbang/https:/gist.github.com/tivrfoa/bb5deb269de39eb8fca9636dd3c9f123#file-gsonhelper-java
java.nio.file.NoSuchFileException: /home/leandro/dev/java/jbang/https:/gist.github.com/tivrfoa/bb5deb269de39eb8fca9636dd3c9f123#file-gsonhelper-java
```
  2. url referencing url with sources and deps
```
leandro:distributions$ ./jbang/bin/jbang --verbose https://gist.github.com/tivrfoa/475c4add1b3df035d13820cb756dacfc
[jbang] Gist url api: https://api.github.com/gists/475c4add1b3df035d13820cb756dacfc
[jbang] Gist url api: https://api.github.com/gists/bb5deb269de39eb8fca9636dd3c9f123
[jbang] [ERROR] Could not download Albums relatively to https://gist.githubusercontent.com/tivrfoa/475c4add1b3df035d13820cb756dacfc/raw/0500521b3906d849293d8893a8ee34e7c76f4209/Main.java
```

Reply:
> But then I found out that if we use history[0].version, then it's the same path for all files! 😃

interesting, do you get the right version of the files then ?

Reply:
> > But then I found out that if we use history[0].version, then it's the same path for all files! 😃
> 
> interesting, do you get the right version of the files then ?

Yes. history[0] is the most recent.
You can check here for example:
https://api.github.com/gists/8e6ea001f168fd4ef14763ceca3e5ab6